core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapYearInComparison.java

```diff
@@ -0,0 +1,233 @@
+            return switch (expression.getOperator()) {
+                case EQUAL -> between(argument, argumentType, calculateRangeStartInclusive(year, argumentType), calculateRangeEndInclusive(year, argumentType));
+                case NOT_EQUAL -> new NotExpression(between(argument, argumentType, calculateRangeStartInclusive(year, argumentType), calculateRangeEndInclusive(year, argumentType)));
+                case IS_DISTINCT_FROM -> or(
+                        new IsNullPredicate(argument),
+                        new NotExpression(between(argument, argumentType, calculateRangeStartInclusive(year, argumentType), calculateRangeEndInclusive(year, argumentType))));
+                case LESS_THAN -> new ComparisonExpression(LESS_THAN, argument, toExpression(calculateRangeStartInclusive(year, argumentType), argumentType));
+                case LESS_THAN_OR_EQUAL -> new ComparisonExpression(LESS_THAN_OR_EQUAL, argument, toExpression(calculateRangeEndInclusive(year, argumentType), argumentType));
+                case GREATER_THAN -> new ComparisonExpression(GREATER_THAN, argument, toExpression(calculateRangeEndInclusive(year, argumentType), argumentType));
+                case GREATER_THAN_OR_EQUAL -> new ComparisonExpression(GREATER_THAN_OR_EQUAL, argument, toExpression(calculateRangeStartInclusive(year, argumentType), argumentType));
+            };
+        }
+
+        private Object calculateRangeStartInclusive(int year, Type type)
+        {
+            if (type == DATE) {
+                LocalDate firstDay = LocalDate.ofYearDay(year, 1);
+                return firstDay.toEpochDay();
+            }
+            if (type instanceof TimestampType timestampType) {
+                if (timestampType.isShort()) {
+                    LocalDateTime dateTime = LocalDateTime.of(year, 1, 1, 0, 0);
+                    return dateTime.toEpochSecond(ZoneOffset.UTC) * MICROSECONDS_PER_SECOND + LongMath.divide(dateTime.getNano(), NANOSECONDS_PER_MICROSECOND, UNNECESSARY);
+                }
+                LocalDateTime dateTime = LocalDateTime.of(year, 1, 1, 0, 0);
+                long endInclusiveMicros = dateTime.toEpochSecond(ZoneOffset.UTC) * MICROSECONDS_PER_SECOND
+                        + LongMath.divide(dateTime.getNano(), NANOSECONDS_PER_MICROSECOND, UNNECESSARY);
+                return new LongTimestamp(endInclusiveMicros, 0);
+            }
+            throw new UnsupportedOperationException("Unsupported type: " + type);
+        }
+
+        private Object calculateRangeEndInclusive(int year, Type type)
+        {
+            if (type == DATE) {
+                LocalDate lastDay = LocalDate.ofYearDay(year, 1).with(lastDayOfYear());
+                return lastDay.toEpochDay();
+            }
+            if (type instanceof TimestampType timestampType) {
+                if (timestampType.isShort()) {
+                    LocalDateTime dateTime = LocalDateTime.MAX.withYear(year).with(lastDayOfYear());
```

Comment:
Is `.with(lastDayOfYear())` still necessary in the context of using `LocalDateTime.MAX` ?